src/Airship/Internal/Route.hs

```diff
@@ -76,40 +149,70 @@ star = Route [RestUnbound]
-    matchesRoute' (ph:ptl)  (Bound sh:stt)  (h, dispatch) True
-        | ph == sh
-                                                    = matchesRoute' ptl stt (h, dispatch ++ [ph]) True
-    matchesRoute' (ph:ptl)  (Bound sh:stt)  (h, dispatch) False
-        | ph == sh
-                                                    = matchesRoute' ptl stt (h, dispatch) False
-    matchesRoute' (ph:ptl)  (Var t:stt)     acc   _ = matchesRoute' ptl stt (insert t ph (fst acc), snd acc) True
-    matchesRoute' _         _               _acc  _ = Nothing
+
+route :: Trie (RouteLeaf a)
+      -> BC8.ByteString
+      -> Maybe (Resource a, (HashMap Text Text, [Text]))
+route routes pInfo = let matchRes = Trie.match routes pInfo
+                     in matchRoute' routes matchRes mempty Nothing
+
+
+matchRoute' :: Trie (RouteLeaf a)
+            -> Maybe (B.ByteString, RouteLeaf a, B.ByteString)
+            -> [Text]
+            -> Maybe B.ByteString
+            -> Maybe (Resource a, (HashMap Text Text, [Text]))
+matchRoute' _routes Nothing _ps _dsp =
+    -- Nothing even partially matched the route
+    Nothing
+matchRoute' routes (Just (matched, RouteMatchOrVar r vars, "")) ps dsp =
+    -- The matched key is also a prefix of other routes, but the
+    -- entire path matched so handle like a RouteMatch.
+    matchRoute' routes (Just (matched, RouteMatch r vars, "")) ps dsp
+matchRoute' _routes (Just (matched, RouteMatch r vars, "")) ps dsp =
+    -- The entire path matched so return the resource, params, and
+    -- dispatch path
+    Just (r, (fromList $ zip vars ps, dispatchList dsp matched))
+    where
+        dispatchList (Just d) m = toTextList $ B.concat [d, m]
+        dispatchList Nothing _ = mempty
+        toTextList bs = (T.pack . BC8.unpack) <$> BC8.split '/' bs
+matchRoute' _routes (Just (_matched, RouteMatch _r _vars, _)) _ps _dsp =
+    -- Part of the request path matched, but the trie value at the
+    -- matched prefix is not an RVar or RouteMatchOrVar so there is no
+    -- match.
+    Nothing
+matchRoute' routes (Just (matched, RouteMatchOrVar _r _vars, rest)) ps dsp =
+    -- Part of the request path matched and the trie value at the
+    -- matched prefix is a RouteMatchOrVar so handle it the same as if
+    -- the value were RVar.
+    matchRoute' routes (Just (matched, RVar, rest)) ps dsp
+matchRoute' routes (Just (matched, RVar, rest)) ps dsp =
+    -- If the rest of the request path does not begin with a '/' then
+    -- the request is for another resource and this is not a match.
+    case (not (BC8.null rest)) && (BC8.head rest == '/') of
+        True ->
+            -- Part of the request path matched and the trie value at the
+            -- matched prefix is a RVar so calculate the key for the next part
+            -- of the route and continue attempting to match.
+            let nextKey = B.concat [ hash $ B.concat [matched, "var"]
+                                   , BC8.dropWhile (/='/') $ BC8.dropWhile (=='/') rest
```

Comment:
This introduces a bit of inconsistency. 

The request `/foo//bar` will not match the route `/foo/bar`. But `/foo//bar` will match the route `/foo/:var`.

It's a separate discussion of whether we should drop repeat or trailing`/`s, but that would break backwards compatibility.